File: kotlin/math/main_1010.kt
import kotlin.math.min

fun main() {
    val dp = Array(31) { size -> IntArray(size / 2 + 1) { 1 } }
    for (i in 2..30){
        for (j in 1..(i / 2)){
            dp[i][j] = (dp[i - 1][j - 1] + dp[i - 1][min(j, i - j - 1)])
        }
    }

    val count = readLine()!!.toInt()
    for (i in 1..count){
        val (k, n) = readLine()!!.split(' ').map { it.toInt() }
        println(dp[n][min(k, n-k)])
    }
}

File: cpp/step/math/main_1010.cpp
#include <iostream>
#include <algorithm>
using namespace std;

int main() {
	int** dp = new int* [31];
	dp[1] = new int[1]{ 1 };
	for (int i = 2; i <= 30; ++i) {
		dp[i] = new int[i / 2 + 1]{ 1, };
		for (int j = 1; j <= i / 2; ++j) {
			dp[i][j] = dp[i - 1][j - 1] + dp[i - 1][min(i - j - 1, j)];
		}
	}
	
	int count;
	cin >> count;
	for (int c = 0; c < count; ++c) {
		int n, k;
		cin >> k >> n;
		cout << dp[n][min(k, n - k)] << '\n';
	}
}